packages/embeds/lib/fetchAndStore.js

```diff
@@ -0,0 +1,167 @@
+const moment = require('moment')
+const { createUrlPrefixer } = require('@orbiting/backend-modules-assets/lib/urlPrefixing')
+const proxyUrl = createUrlPrefixer()
+const Redlock = require('redlock')
+const debug = require('debug')('embeds')
+const twitter = require('./twitter')
+const linkPreview = require('./linkPreview')
+const types = {
+  TwitterEmbed: {
+    ...twitter
+  },
+  LinkPreview: {
+    ...linkPreview
+  }
+}
+
+const { REQUEST_TIMEOUT_SECS } = linkPreview
+
+const TTL_DB_ENTRIES_SECS = 6 * 60 * 60 // 6h
+const MAX_REQUEST_URL_EACH_SECS = 30
+
+const LOCK_RETRY_DELAY_MS = 600
+const LOCK_RETRY_COUNT = Math.ceil(
+  3 * ((REQUEST_TIMEOUT_SECS * 1000) / LOCK_RETRY_DELAY_MS)
+)
+const LOCK_TTL_MS = LOCK_RETRY_COUNT * LOCK_RETRY_DELAY_MS
+
+const redlock = (redis) => {
+  return new Redlock(
+    [redis],
+    {
+      retryCount: LOCK_RETRY_COUNT,
+      retryDelay: LOCK_RETRY_DELAY_MS
+    }
+  )
+}
+
+const fetchContent = async (url) => {
+  if (twitter.REGEX.test(url)) {
+    const tweetId = twitter.REGEX.exec(url)[1]
+    return {
+      type: 'TwitterEmbed',
+      content: await twitter.getTweetById(tweetId)
+    }
+  }
+  return {
+    type: 'LinkPreview',
+    content: await linkPreview.getLinkPreviewByUrl(url)
+  }
+}
+
+const proxyContent = (content, type) => {
+  const newContent = {}
+  for (const key of types[type].imageKeys) {
+    newContent[key] = proxyUrl(content[key])
+  }
+  return {
+    ...content,
+    ...newContent
+  }
+}
+
+const fetchEmbed = async ({ url, doUpdate = false }, context) => {
+  debug(
+    'fetchLinkPreview url: %s doUpdate: %s',
+    url,
+    doUpdate
+  )
+
+  const {
+    pgdb,
+    redis,
+    loaders
+  } = context
+
+  const redisKey = `embeds:fetch:${url.toLowerCase()}`
+  const lockKey = `locks:${redisKey}`
+
+  const lock = await redlock(redis).lock(lockKey, LOCK_TTL_MS)
+  try {
+    const existingEmbed = await loaders.Embed.byUrl.load(url)
+    if (!doUpdate && existingEmbed) {
+      return existingEmbed
+    }
+
+    // throttle
+    const throttled = await redis.getAsync(redisKey)
+    if (throttled) {
+      debug('throttle prevented url from beeing fetched %s', url)
+      return null
+    }
+    await redis.setAsync(
+      redisKey,
+      1,
+      'EX', MAX_REQUEST_URL_EACH_SECS
+    )
+
+    const { type, content } = await fetchContent(url)
+
+    if (content) {
+      let dbEntry
+      if (!existingEmbed) {
+        dbEntry = await pgdb.public.embeds.insertAndGet({
+          url,
+          host: new URL(url).hostname,
+          type,
+          typeId: content.id,
```

Comment:
Some ideas here, as `typeId` feels a bit of considering ID refers to content and not an ID to some sort of "type".

- normalize `content` to contain `id` and index `content->>id`
- rename field to `contentId`

Reply:
renamed to `contentId`